File: Yona/Yona/Core/LoadState.swift
//
//  LoadState.swift
//  Yona
//
//  One enum that every screen uses to render loading / empty / error / content,
//  so "online-only" never means a silent blank screen.
//

import Foundation

enum LoadState<Value> {
    case idle
    case loading
    case loaded(Value)
    case failed(Error)

    var value: Value? {
        if case let .loaded(value) = self { return value }
        return nil
    }

    var isLoading: Bool {
        if case .loading = self { return true }
        return false
    }

    var error: Error? {
        if case let .failed(error) = self { return error }
        return nil
    }
}

File: android/app/src/main/java/com/yona/app/core/LoadState.kt
package com.yona.app.core

/**
 * The four explicit states every async screen renders, mirroring the iOS
 * `LoadState<T>` enum: idle, loading, loaded(value), or failed(message).
 */
sealed interface LoadState<out T> {
    data object Idle : LoadState<Nothing>
    data object Loading : LoadState<Nothing>
    data class Loaded<T>(val value: T) : LoadState<T>
    data class Failed(val message: String) : LoadState<Nothing>
}
